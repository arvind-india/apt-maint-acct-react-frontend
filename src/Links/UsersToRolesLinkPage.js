import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import {
    Input,
    Label
} from 'reactstrap'

import {
  MdHome,
  MdThumbUp,
  MdThumbDown
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { FlashMessage } from '../_components'
import { userActions, roleActions } from '../_actions'

let module = 'users-roles' // module name


export class UsersToRolesLink extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      selectedOptionInLeftList: '',
      selectedOptionsInDList: [],
      selectedOptionsInAList: []
    }
    this.handleChangeInLeftList = this.handleChangeInLeftList.bind(this)
    this.handleChangeInDetachedList = this.handleChangeInDetachedList.bind(this)
    this.attachItems = this.attachItems.bind(this)
    this.handleChangeInAttachedList = this.handleChangeInAttachedList.bind(this)
    this.detachItems = this.detachItems.bind(this)
  }

  componentDidMount() {
    this.props.getAllUsers()
    this.props.getAllRoles()
  }

  render() {
    const { alert } = this.props
    return (
      <div>
        <h3>Users To Roles Link</h3>
        {alert.message && <FlashMessage text={alert.message} delay={2100}/>}
        {this.show()}
      </div>
    )
  }

  show() {
    return <div className="list-to-list">
      { this.showLeftLabel() }
      { this.showLeftInput() }
      { this.showLeftButton() }

      { this.showAttachedLabel() }
      { this.showAttachedInput() }
      { this.showDetachButton() }

      { this.showDetachedLabel() }
      { this.showDetachedInput() }
      { this.showAttachButton() }
    </div>
  }

  showLeftLabel() {
    return <Label
      for="leftItem"
      className="llabel"
      >Select an User</Label>
  }
  showLeftInput() {
    const { users } = this.props
    return <Input
      id="leftItem"
      type="select"
      name="leftItem"
      size="20"
      className="lselect"
      onChange={this.handleChangeInLeftList}
    >
    { users.items && users.items.map(each =>
      <option value={each.id} title={each.email} key={each.id}
        >{each.first_name?each.first_name:each.name}</option>)
    }
    </Input>
  }
  handleChangeInLeftList(event) {
    const { value } = event.target
    this.setState({
      selectedOptionInLeftList: value
    })

    this.props.getMyRoles(value)
  }
  showLeftButton() {
    return <div class="homeButton">
      <Button
        color="link"
      ><Link to="/home"><MdHome/> Home</Link></Button>
    </div>
  }

  showAttachedLabel() {
    return <Label
      for="attachedItems"
      className="alabel"
    >Assigned Roles</Label>
  }
  showAttachedInput() {
    const { usersToRoles } = this.props
    const { selectedOptionInLeftList } = this.state

    return <Input
      id="attachedItems"
      type="select"
      name="attachedItems"
      size="20"
      className="aselect"
      onChange={this.handleChangeInAttachedList}
      multiple
    >
    { selectedOptionInLeftList &&
      usersToRoles.items &&
      usersToRoles.items.map(each =>
      <option value={each.id} title={each.description} key={each.id}
        >{each.name} {each.inherits?' (inherits '+each.inherits+')':''}</option>)
    }
    </Input>
  }
  handleChangeInAttachedList(event) {
    const { selectedOptions } = event.target
    this.setState({
      selectedOptionsInAList: Array.from(selectedOptions)
    })
  }
  showDetachButton() {
    const { authzn } = this.props
    const { selectedOptionsInAList } = this.state

    return <Button
      id="detachButton"
      color="danger"
      className="dbutton"
      onClick={this.detachItems}
      disabled={selectedOptionsInAList.length === 0}
      hidden={!authzn.allowsEdit}
      title="Unassign selected roles"
    ><MdThumbDown/> Unassign</Button>
  }

  detachItems() {
    const { usersToRoles } = this.props
    const { selectedOptionsInAList, selectedOptionInLeftList } = this.state

    let id = selectedOptionInLeftList
    let ids_toBeRemoved = selectedOptionsInAList.map(e => e.value)
    let ids = usersToRoles.items.map(e => e.id.toString())
    let ids_toBeRetained = ids.filter(each => !ids_toBeRemoved.includes(each))

    this.props.updateMyRoles(id, ids_toBeRetained)
    this.setState({ selectedOptionsInAList: [] })
  }

  showDetachedLabel() {
    return <Label
      for="detachedItems"
      className="dlabel"
    >Available Roles</Label>
  }
  showDetachedInput() {
    const { usersToRoles, roles } = this.props
    const { selectedOptionInLeftList } = this.state
    let available = [];
    let grantedIDs = []
    if(usersToRoles.items) {
      grantedIDs = usersToRoles.items.map(each => each.id)
    }
    if(roles.items && selectedOptionInLeftList) {
      available = grantedIDs.length ?
        roles.items.filter(each => !grantedIDs.includes(each.id)) :
        roles.items
    }
    return <Input
      id="detachedItems"
      type="select"
      name="detachedItems"
      size="20"
      className="dselect"
      onChange={this.handleChangeInDetachedList}
      multiple
    >
    {
      available.map(each =>
      <option value={each.id} title={each.description} key={each.id}
        >{each.name} {each.inherits?' (inherits '+each.inherits+')':''}</option>)
    }
    </Input>
  }
  handleChangeInDetachedList(event){
    const { selectedOptions } = event.target
    this.setState({
      selectedOptionsInDList: Array.from(selectedOptions)
    })
  }
  showAttachButton() {
    const { authzn } = this.props
    const { selectedOptionsInDList } = this.state

    return <Button
      id="attachButton"
      color="success"
      className="abutton"
      onClick={this.attachItems}
      disabled={selectedOptionsInDList.length === 0}
      hidden={!authzn.allowsEdit}
      title="Assign selected roles"
    ><MdThumbUp/> Assign</Button>
  }

  attachItems() {
    const { usersToRoles } = this.props
    const { selectedOptionsInDList, selectedOptionInLeftList } = this.state

    let id = selectedOptionInLeftList
    let ids = usersToRoles.items.map(e => e.id)
    let ids_toBeAdded = selectedOptionsInDList.map(e => e.value)
    let new_ids = ids.concat(ids_toBeAdded)

    this.props.updateMyRoles(id, new_ids)
    this.setState({ selectedOptionsInDList: [] })
  }
}

function mapStateToProps(state) {
  const { alert, users, usersToRoles, roles, authorizations } = state
  const authzn = authorizations[module]
  return {
    alert,
    users,
    usersToRoles,
    roles,
    authzn
  }
}

function mapDispatchToProps(dispatch) {
  return {
    getAllRoles: () => {
      dispatch(roleActions.getAll())
    },
    getAllUsers: () => {
      dispatch(userActions.getAll())
    },
    getMyRoles: (value) => {
      dispatch(userActions.getMyRoles(value))
    },
    updateMyRoles: (id, ids) => {
      dispatch(userActions.updateMyRoles(id, ids))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(UsersToRolesLink)
