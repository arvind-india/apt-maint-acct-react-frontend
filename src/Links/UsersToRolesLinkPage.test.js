// UsersToRolesLinkPage.test.js0]

import React from 'react'
import { shallow } from 'enzyme'

import { UsersToRolesLink } from './UsersToRolesLinkPage'

const mockUsers = [
  { id: 1, name: '0', first_name: 'G1', email: 'test on G1' },
  { id: 2, name: '0', first_name: 'G2', email: 'test on G2' }
]

const mockRoles = [
  { id: 1, name: 'user1', inherits: 'test', description: 'test user1' },
  { id: 2, name: 'user2', inherits: 'test', description: 'test user2' }
]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    users: { items: mockUsers },
    usersToRoles: { items: [] },
    roles: { items: mockRoles },
    getAllUsers: jest.fn(),
    getAllRoles: jest.fn(),
    getMyRoles: jest.fn(),
    updateMyRoles: jest.fn()
  }
]

const component = shallow(<UsersToRolesLink {...mockProps[0]} />)

describe('UsersToRolesLinkPage params test', () => {
  it('should render UsersPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionInLeftList')).toBeDefined()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionsInDList')).toBeDefined()
  })
  it('should have model state', () => {
    expect(component.state('selectedOptionsInAList')).toBeDefined()
  })
})


const selectedOptions = [mockRoles[0].id, mockRoles[1].id]

describe('UsersToRolesLinkPage test: select a user and select two roles', () => {

  component.find('#leftItem').simulate('change', {target: {name:'id', value: mockUsers[0].id}})

  it('should check options in Left List', () => {
    expect(component.state('selectedOptionInLeftList')).toEqual(mockUsers[0].id)
  })

  component
    .find('#detachedItems')
    .simulate('change', { target: {name: 'id', selectedOptions: selectedOptions} })

  it('should check options in detached list', () => {
    expect(component.state('selectedOptionsInDList'))
      .toEqual(selectedOptions)
  })

  component
    .find('#attachedItems')
    .simulate('change', { target: {name: 'id', selectedOptions: selectedOptions} })

  it('should have two options in the attached list', () => {
    expect(component.state('selectedOptionsInAList'))
      .toEqual(selectedOptions)
  })

})



const component2 = shallow(<UsersToRolesLink {...mockProps[0]} />)

describe('UsersToRolesLinkPage test: Attach selected items', () => {

  component2
    .find('#leftItem')
    .simulate('change', {target: {name:'id', value: mockUsers[0].id} })

  component2
    .find('#detachedItems')
    .simulate('change', {target: {name: 'id', selectedOptions: selectedOptions} })

  component2
    .find('Button.abutton')
    .simulate('click')

  it('should now remove two options from detached list', () => {
    expect(component2.state('selectedOptionsInDList')).toEqual([])
  })

})



const component3 = shallow(<UsersToRolesLink {...mockProps[0]} />)

describe('UsersToRolesLinkPage test: Detach selected items', () => {

  component3
    .find('#leftItem')
    .simulate('change', { target: {name:'id', value: mockUsers[0].id} })

  component3
    .find('#attachedItems')
    .simulate('change', { target: {name: 'id', selectedOptions: selectedOptions} })

  component3
    .find('Button.dbutton')
    .simulate('click')

  it('should now remove two options from attached list', () => {
    expect(component3.state('selectedOptionsInAList')).toEqual([])
  })

})
