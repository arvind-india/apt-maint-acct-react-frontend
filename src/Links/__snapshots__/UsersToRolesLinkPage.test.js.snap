// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersToRolesLinkPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Users To Roles Link
    </h3>
    <FlashMessage
        delay={2100}
        text="testingAlert"
    />
    <div
        className="list-to-list"
    >
        <Label
            className="llabel"
            for="leftItem"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Select an User
        </Label>
        <Input
            className="lselect"
            id="leftItem"
            name="leftItem"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            <option
                title="test on G1"
                value={1}
            >
                G1
            </option>
            <option
                title="test on G2"
                value={2}
            >
                G2
            </option>
        </Input>
        <div
            class="homeButton"
        >
            <Button
                color="link"
                tag="button"
            >
                <Link
                    replace={false}
                    to="/home"
                >
                    <MdHome />
                     Home
                </Link>
            </Button>
        </div>
        <Label
            className="alabel"
            for="attachedItems"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Assigned Roles
        </Label>
        <Input
            className="aselect"
            id="attachedItems"
            multiple={true}
            name="attachedItems"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            
        </Input>
        <Button
            className="dbutton"
            color="danger"
            disabled={false}
            hidden={false}
            id="detachButton"
            onClick={[Function]}
            tag="button"
            title="Unassign selected roles"
        >
            <MdThumbDown />
             Unassign
        </Button>
        <Label
            className="dlabel"
            for="detachedItems"
            tag="label"
            widths={
                Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
            }
        >
            Available Roles
        </Label>
        <Input
            className="dselect"
            id="detachedItems"
            multiple={true}
            name="detachedItems"
            onChange={[Function]}
            size="20"
            tag="p"
            type="select"
        >
            <option
                title="test user1"
                value={1}
            >
                user1
                 
                 (inherits test)
            </option>
            <option
                title="test user2"
                value={2}
            >
                user2
                 
                 (inherits test)
            </option>
        </Input>
        <Button
            className="abutton"
            color="success"
            disabled={false}
            hidden={false}
            id="attachButton"
            onClick={[Function]}
            tag="button"
            title="Assign selected roles"
        >
            <MdThumbUp />
             Assign
        </Button>
    </div>
</div>,
]
`;
